tempoclock sync

telephon opened the pull request:
Method thanks to @redFrik.

brianlheim commented on SCClassLibrary/Common/Core/Clock.sc:
```diff
@@ -327,6 +327,48 @@ elapsed time is whatever the system clock says it is right now. elapsed time is
+			this.sched(next, { this.tempo_(tempo); nil });
+		} {							// else interpolate
+			this.sched(next, {		// start at next beat
+				durCur = this.tempo.reciprocal;
```
These variables need clearer names, and I think this block could definitely do with a good once-over to make sure all these variables are really necessary and to get the code in a more readable state. For instance, `delta` is assigned 12 lines before it's used, and interrupts the flow of calculations. It also looks like `delta`, `factor`, and `stepSize` can probably be eliminated since they just store the result of small math ops and are only used once or twice after that.

In addition, I would recommend adding parentheses to clearly show order of operations, especially when relying on SC's lack of operator precedence (lines 345, 350, 351).

brianlheim commented on SCClassLibrary/Common/Core/Clock.sc:
```diff
@@ -327,6 +327,48 @@ elapsed time is whatever the system clock says it is right now. elapsed time is
 		^this.beats - this.bars2beats(this.bar)
 	}
 
+	sync { arg tempo, secs = 4, resolution = 1;
```
Also, please add a brief method description above the definition.

brianlheim commented on HelpSource/Classes/TempoClock.schelp:
```diff
@@ -247,6 +247,38 @@ t.beats2secs(0) // how many seconds after startup did beat 0 occur?
 Converts absolute strong::seconds:: to absolute beats. Only works for times in the current tempo. If the tempo changes any computed time in future will be wrong.
 
+method::sync
+Gradually synchronize one or more clocks to the same downbeat and tempo.
```
This method only synchronizes one clock (the current one), so this is a misleading statement.

telephon commented:
I'm not completely convinced about the name `fadeTempo`, because the method fades to both beat and tempo. Any suggestions welcome!

brianlheim commented:
@telephon These changes don't address most of my code review. Is that intentional?

redFrik commented:
fadeTempo sounds strange to me.  beatMatch or perhaps tempoMatch?

> 22 mar 2017 kl. 22:12 skrev Julian Rohrhuber <notifications@github.com>:
> 
> I'm not completely convinced about the name fadeTempo, because the method fades to both beat and tempo. Any suggestions welcome!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 


  #|
     fredrikolofsson.com     musicalfieldsforever.com
  |#

telephon commented:
@brianlheim I would never intentionally ignore your code review! :) I had assumed I had covered them all, but looking through, there are a few that still need to be done. No time now, unfortunately...

brianlheim commented:
@telephon OK, no worries! I was just a little confused. I wouldn't mind working on it a bit if you don't mind.

telephon commented:
yes, why not - please go ahead!

telephon commented:
@brianlheim would you still like to do it, or should we try and fix it together?

brianlheim commented:
@telephon I don't think I'll have time to work on it this week... I've been knee-deep in fixing #2441.

crucialfelix commented:
`blendToTempo` - I find this easily understandable.

`syncTempo` means sync to something else. but what else is there that it is syncing too ? It suggests an other but there is no other.

`tweenTempo` - this is the old flash terminology. 

`transitionTempoTo` - current css animation terminology

But rather than stick this whole thing on Tempo, why don't you make it a separate service that takes a getter and a setter. On each step it gets the value, interpolates and then calls the setter. 

Then Tempo can call that service to perform the blend/transition/animation and supply it with the getter and setter.

You can just as easily interpolate multiple targets at the same time. You could return a function that allows cancelling the operation while it is in motion.

You could supply a tweening / easing function like "bouncing ball".

crucialfelix commented:
Definitely should be parameterized with an easing function. This is too much code just to perform one simple linear fade.

snappizz commented:
This is cool. Is there any way to unit test clock functionality?

redFrik commented:
please don't call it blend or fade.  it does a bit more than simple interpolation - it guarantees a downbeat in a new tempo at some future time.

this method used on its own is not so interesting.  the purpose is rather to make a tempoclock play in time with some other music happening.  this can be another clock in sc, sound coming from some other application or music playing acoustically in the room.  there's your reference.

so in my opinion better names would hint at adjusting, synchronizing or beat matching.

there's a quark for tweening... https://github.com/supercollider-quarks/Ease
and you can use it to interpolate a clock's tempo...
{|t| EaseInAtan.(t/99)}.dup(100).plot
c= TempoClock(127/60);
p= Pbind().play(c);
Routine.run({100.do{|t| c.tempo= EaseInAtan.(t/99).range(127/60, 160/60); 0.01.wait}});
Routine.run({100.do{|t| c.tempo= EaseInAtan.(t/99).range(160/60, 80/60); 0.01.wait}});
the problem though is that this will ignore beats and bars.  but i could imagine trying to code a variant of this quark specially for tempoclock.  and then we wouldn't need to include this sync method in the main distro.
_f

> 1 maj 2017 kl. 17:13 skrev Chris Sattinger <notifications@github.com>:
> 
> blendToTempo - I find this easily understandable.
> 
> syncTempo means sync to something else. but what else is there that it is syncing too ? It suggests an other but there is no other.
> 
> tweenTempo - this is the old flash terminology.
> 
> transitionTempoTo - current css animation terminology
> 
> But rather than stick this whole thing on Tempo, why don't you make it a separate service that takes a getter and a setter. On each step it gets the value, interpolates and then calls the setter.
> 
> Then Tempo can call that service to perform the blend/transition/animation and supply it with the getter and setter.
> 
> You can just as easily interpolate multiple targets at the same time. You could return a function that allows cancelling the operation while it is in motion.
> 
> You could supply a tweening / easing function like "bouncing ball".
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 


  #|
     fredrikolofsson.com     musicalfieldsforever.com
  |#

telephon commented:
I think it is useful to have one simple method in the main distro. If we want to pass in some kindof warp or ease function, we need to calculate the integration steps of the procedure so that it ends up on the right beat.

crucialfelix commented:
Ah, so in `S` seconds it will be playing at tempo `T` and have a downbeat at that time. 

This assumes that the Other will also have a downbeat at that time. So you've calculated how many seconds till that next downbeat (given the tempo and current beat of Other).

`convergeToTempo` ?

I guess sync makes sense. It just seemed strange because with only tempo and seconds it seems to just be interpolating an automation of tempo.

telephon commented on SCClassLibrary/Common/Core/Clock.sc:
```diff
@@ -327,6 +327,48 @@ elapsed time is whatever the system clock says it is right now. elapsed time is
+		time = secs - (this.tempo.reciprocal * next);
+		if(time < next) {						// jump directly
+			this.tempo_(next / secs);		// set a high tempo
+			this.sched(next, { this.tempo_(tempo); nil });
```
what do you mean?

telephon commented on SCClassLibrary/Common/Core/Clock.sc:
```diff
@@ -327,6 +327,48 @@ elapsed time is whatever the system clock says it is right now. elapsed time is
+					if(index < steps) {
+						durCur = durCur + stepSize;
+						tmp = (durCur * factor).reciprocal;
+						this.tempo_(tmp);
```
no this isn't possible because the line above changes `durCur` in each iteration.

telephon commented:
I've just changed the name to `syncToBeat` -- it seems that this is the clearest, in particular given the misunderstandings so far. Also, I've changed the argument order: seconds first, then tempo. You may simply want to reach a beat after n seconds, and not change the tempo.

If you all prefer `convergeToTempo` or something else, I'm fine to change it once more.

redFrik commented:
this is going nowhere - just misunderstandings and naming discussions. let's forget about it and not add .sync to the main library.  it's fine to have it available as an extension on the swiki (as it has been since 2005).

brianlheim commented:
No worries, @redFrik, sorry that this couldn't move fast enough for you.

I would invite anyone who is frustrated with the speed of discussions on GitHub to come to the weekly dev meetings, which are naturally a more directed and productive environment for discussions like these.

telephon commented:
well I'd aporeciate if the work we all invested wouldn't be completely wasted nevertheless

telephon commented:
I can close it again if we agree that this goes nowhere. I had the impression we are almost there.

brianlheim commented:
- Pick a clear name and argue for it
- Why is the phase argument named "quant"? I don't think this makes sense.
- "seconds" and "resolution" arguments do not need to be arbitrarily limited. IMO it is much better to document practical limitations than to enforce them.

telephon commented:
- `syncToBeat` ssufficiently says what it does and why.

- quant is the way to specify a next possible phase in clocks (see quant help file)

- good point, will do that.

brianlheim commented:
> quant is the way to specify a next possible phase in clocks (see quant help file)

So from what I'm understanding, it's actually the number of beats, offset from the new tempo, minus one, at which the clock will sync? So if quant was 2.0, it would actually sync at `seconds + (1.0 * tempo)`? I'm mainly trying to figure out why the arg is named "quant" but the description says "[clock] phase." The documentation for Quant is very confusing at a first glance...

brianlheim commented:
And FWIW, i like the name syncToBeat. the fact that it's interpolating is implied by the parameter list.

telephon commented:
re quant: The `seconds` argument tells you when the final state will be reached. The `tempo` tells you what tempo it will have then. The `quant` tells you in what clock phase it will be. This clock phase is specified by an integer number of beats, and optionally by an offset relative to it. You are right it is a little hard to explain, maybe we find the right words for the quant helpfile.

brianlheim commented:
Ah okay, that makes much more sense. Thank you for explaining!

brianlheim commented:
This is so so close to being done, would you mind resolving this last thing @telephon ?

> "seconds" and "resolution" arguments do not need to be arbitrarily limited. IMO it is much better to document practical limitations than to enforce them.

> good point, will do that.

telephon commented:
I've been having an email conversation with Fredrik, and I tend to agree with him that the method needs not be added in the current form. I'd like to think a bit more about it and see if I can simplify+generalize so that it is better justified.

brianlheim commented:
Ok. I am going to close this then. A new PR can be opened where there's something concrete to discuss.

telephon commented:
thanks! could you maybe add it to the PR's that are there to be reopened later?

LFSaw commented:
(closed after discussion on [sc-dev](http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/PR-backlog-td7632678.html) and closing documented on [projects/7](https://github.com/supercollider/supercollider/projects/7))
